### PR TITLE
MAINT: stats: bootstrap: fix error when bootstrap distribution is degenerate

### LINKED ISSUE #14804: BUG: scipy.stats.bootstrap unable to deal with constant samples
### Describe your issue.

It seems scipy.stats.boostrap will raise an error when the sample data are constant numbers. From the statistical perspective, I agree that bootstrap may not be used on constant numbers. So raising an error may be ok, but we may also consider the confidence interval to be 0 which would not stop the program. 

If you think it's a better idea to raise an error, maybe the error message could be more self-explain ? 

Thanks

### Reproducing Code Example

```python
from scipy import stats
import numpy as np
stats.bootstrap([[1]*10],statistic=np.mean)
```

### Error message

```shell
Traceback (most recent call last):
  File "/Users/saibo/Research/project001/BT-eval-metric-corr/debug.py", line 24, in <module>
    stats.bootstrap([[1]*10],statistic=np.mean)
  File "/opt/anaconda3/envs/BT-eval-metric/lib/python3.8/site-packages/scipy/stats/_bootstrap.py", line 454, in bootstrap
    ci_l = percentile_fun(theta_hat_b, interval[0]*100)
  File "/opt/anaconda3/envs/BT-eval-metric/lib/python3.8/site-packages/scipy/stats/_bootstrap.py", line 78, in _percentile_along_axis
    percentiles[indices] = np.percentile(theta_hat_b_i, alpha_i)
  File "<__array_function__ internals>", line 5, in percentile
  File "/opt/anaconda3/envs/BT-eval-metric/lib/python3.8/site-packages/numpy/lib/function_base.py", line 3818, in percentile
    return _quantile_unchecked(
  File "/opt/anaconda3/envs/BT-eval-metric/lib/python3.8/site-packages/numpy/lib/function_base.py", line 3937, in _quantile_unchecked
    r, k = _ureduce(a, func=_quantile_ureduce_func, q=q, axis=axis, out=out,
  File "/opt/anaconda3/envs/BT-eval-metric/lib/python3.8/site-packages/numpy/lib/function_base.py", line 3515, in _ureduce
    r = func(a, **kwargs)
  File "/opt/anaconda3/envs/BT-eval-metric/lib/python3.8/site-packages/numpy/lib/function_base.py", line 4047, in _quantile_ureduce_func
    ap.partition(concatenate((
ValueError: kth(=-9223372036854765809) out of bounds (9999)
```

### SciPy/NumPy/Python version information

1.7.1 1.20.3 sys.version_info(major=3, minor=8, micro=10, releaselevel='final', serial=0)

### PR DESCRIPTION
**Pull request by mdhaber**
#### What does this implement/fix?
In master:
```python3
import numpy as np
from scipy.stats import bootstrap
n = 35
data = n * [10000.] # + np.random.rand(n)
print(bootstrap([data,], np.mean, method='bca'))  # ValueError: kth(=-9223372036854765809) out of bounds (9999)
```

When the bootstrap distribution is degenerate (e.g. has only one unique value), the BCa formulas result in NaNs for the confidence interval alphas (due to 0/0 or 0*-inf). This PR ensures that those NaNs propagate rather than raising an unintelligible error.  Now the result of the calculation above is:
```
BootstrapResult(confidence_interval=ConfidenceInterval(low=nan, high=nan), standard_error=0.0)
```

#### Additional information
Alternatives considered:

The `'percentile'` and `'basic'` confidence intervals for the example above are:
```
BootstrapResult(confidence_interval=ConfidenceInterval(low=10000.0, high=10000.0), standard_error=0.0)
```
This suggests that we could somehow detect the degenerate bootstrap distribution and return this "correct" answer. I'm not sure if that's appropriate, because if the method is BCa, that should be respected, and the BCa formulas (as they are usually written) seem to result in NaNs. (Perhaps they can be rearranged to avoid it, but it's not obvious to me how to do that.)

Another alternative would be to raise an error. However, when the samples are N-dimenensional and NaNs do not occur in all axis-slices, the presence of the NaNs should not stop the rest of the calculations. We could raise an error for 1D samples and when all axis-slices are NaN, but I disagree with that sort of inconsistent behavior.

If we want to emit a warning message, that's fine with me, but I'll leave to another maintainer to recommend. 

Note that I'm not aware of a situation in which this would occur in practice. I've only seen it when the input array is all constant values or there are two or fewer unique elements in the bootstrap distribution (e.g. very few unique values in data or very small `n_resamples`) . @alexsherstinsky suggested this example when looking for edge cases.

### CODE REVIEW

**Comment by alexsherstinsky** [comment]
@mdhaber We are dealing here with an undefined skewness (i.e., the 0/0 situation), correct?  If so, could we potentially rationalize a way to disambiguate it as follows:
* if the degenerate distribution, with the mean subtracted, were to be perturbed by an infinitesimal amount of Gaussian noise with the zero mean and an infinitesimal variance, then the resulting signal would have a zero skewness.
* In other words, if the 3rd moment is 0 (numerator), we can ignore the 2nd moment (in the denominator).

If this is reasonable, then I would prefer this direction (which, using our example, would yield both low=10000.0 and high=10000.0, and std=0.0) to the one that will return NaNs to the user if the user's sample happens have the same value for all data points.

Thank you.

**Comment by mdhaber** [comment]
That's one reason. But part of the purpose of the BCa is to correct for skewness in the bootstrap distribution, right? Why do you get to choose to perturb with Gaussian noise, then?

**Comment by alexsherstinsky** [comment]
> That's one reason. But part of the purpose of the BCa is to correct for skewness in the bootstrap distribution, right? Does the theory still suggest that it is appropriate to perturb with Gaussian noise? (I don't have a lot of familiarity with the theory here, but in general, undefined skewness due to numerically constant input really is undefined, I think.)
> 
> The other things that would need to be resolved come from:
> 
> ```python
>     percentile = _percentile_of_score(theta_hat_b, theta_hat, axis=-1)
>     z0_hat = ndtri(percentile)
> ```
> 
> `theta_hat_b` is the bootstrap distribution and `theta_hat` is the observed value of the statistic.
> 
> * What is the percentile of the score of the observed value of the statistic `theta_hat_b` when it is the same as all values in `theta_hat_b`?
> * When `theta_hat` is distinct from and less than all values in `theta_hat_b`, the percentile of the score is zero. (Or is there some argument for it not being precisely zero?) In that case `ndtri` returns `-inf`. Later, even if `a_hat` is zero as argued above, `a_hat*num1` would become NaN (`0 * -np.inf`).
> 
> I would prefer that `bootstrap` return the correct answer here instead of NaN. (You originally suggested that we raise an error here, right? I replied that it would be better if this could return the correct answer, which I assumed was 10000., 10000. at the time.) But we would need to get some agreement between experts on what is correct. Until then, I thought NaN would be better than my assumption.

@mdhaber Upon further thought and reading through `https://statistics.stanford.edu/sites/g/files/sbiybj6031/f/EFS%20NSF%20226.pdf` last night (albeit rather quickly; could use a more thorough study), I feel that your approach of outputting the `NaN` values is more consistent -- as long as there is an accompanying comment explaining that encountering the `NaN` values is indicative of issues with the data sample supplied.

If the user gets back `NaN` values from `bootstrap()`, then the user can do something like this:
```
        if np.all(np.isclose(metric_values, metric_values[0])):
            # Computation is unnecessary if distribution is degenerate.
            confidence_interval = ConfidenceInterval(
                low=metric_values[0], high=metric_values[0]
            )
```
which would produce the same results as the `basic` method, but not incur unnecessary computation -- this agrees with your point in our email thread.  For more complex data sets, the user can do an upfront validation using one of the aggregate assertions from `https://docs.greatexpectations.io/en/latest/reference/glossary_of_expectations.html` to make sure that the distribution is not degenerate (or whatever they are expecting it to be) and not use `bootstrap()`, unless warranted.

Thank you.

**Comment by mdhaber** [comment]
I added a note about NaNs in the confidence interval.

**Comment by alexsherstinsky** [comment]
🙇

**Comment by mdhaber** [comment]
Is that good or bad?

**Comment by alexsherstinsky** [comment]
> Is that good or bad?

That's a "thank you" (man bowing). :-)

**Comment by mdhaber** [comment]
Thanks. If it looks good to you, please note your opinion by creating a review and selecting the "approve" radio button.

**Comment by alexsherstinsky** [comment]
> Thanks. If it looks good to you, please note your opinion by creating a review and selecting the "approve" radio button.

OK doing that now.  I was not sure what the process was (since I was not formally invited via GitHub to review this PR).  Thanks!

**Review by alexsherstinsky — APPROVED** [review]
This approach is consistent with the theory (as presented in `https://statistics.stanford.edu/sites/g/files/sbiybj6031/f/EFS%20NSF%20226.pdf`).  In order to prevent the occurrence of `NaN` values in the output, users can perform a check, such as:
```
        if np.all(np.isclose(data, data[0])):
            # Computation is unnecessary if distribution is degenerate.
            confidence_interval = ConfidenceInterval(
                low=data[0], high=data[0]
            )
```
Among other benefits, this will avoid an unnecessary expenditure of computational resources.

Thank you.

**Comment by mdhaber** [comment]
Thanks @alexsherstinsky. Yeah, GitHub only seems to allow me to request a review from maintainers and members of the triage team, but I think it is useful for everyone to use the review radio buttons to summarize their opinion of the PR status.

**Review by tirthasheshpatel — COMMENTED** [review]
Sorry, @mdhaber for coming to this soo late! The tests and notes in the documentation look good. I think a warning will be helpful. [`pearsonr`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.PearsonRConstantInputWarning.html#scipy.stats.PearsonRConstantInputWarning) and [`spearmanr`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.SpearmanRConstantInputWarning.html#scipy.stats.SpearmanRConstantInputWarning) throw warnings for constant input cases.

**Comment by mdhaber** [comment]
Thanks @tirthasheshpatel. I'm afraid I won't get back to this immediately, but if you're interested in pushing the change to get it merged, I don't object.

**Comment by tirthasheshpatel** [comment]
I have added a warning `BootstrapDegenerateDistributionWarning` which is emited when `BDa` is used and the bootstrap distribution is degerate:

```python
>>> import numpy as np
>>> from scipy.stats import bootstrap
>>> bootstrap([35*[10000.]], np.mean, method='BCa')
/home/tirthasheshpatel/Desktop/scipy_source/scipy/stats/_bootstrap.py:119: RuntimeWarning: invalid value encountered in double_scalars
  a_hat = num / den
/home/tirthasheshpatel/Desktop/scipy_source/scipy/stats/_bootstrap.py:93: BootstrapDegenerateDistributionWarning: The bootstrap distribution is degenerate; the confidence interval is not defined.
  warnings.warn(BootstrapDegenerateDistributionWarning())
BootstrapResult(confidence_interval=ConfidenceInterval(low=nan, high=nan), standard_error=0.0)
```

Let me know if it looks good. I will merge when you are happy with it.

**Comment by tirthasheshpatel** [comment]
Failures are unrelated. xref #14974

**Comment by mdhaber** [comment]
Thanks @tirthasheshpatel. Since warnings with `nan`s is not really standardized, could we get a third opinion? I feel like the default is to not have a warning, and a warning can always be added but is a bit harder to take away. Is the intent to add warnings whenever we emit nans?

**Comment by tirthasheshpatel** [comment]
> Since warnings with `nan`s is not really standardized, could we get a third opinion? I feel like the default is to not have a warning, and a warning can always be added but is a bit harder to take away. Is the intent to add warnings whenever we emit nans?

I think when we fill in nans implicitly for special cases (like constant input), it's better to emit a warning (to let the user know that something is wrong, hence, we have output nan). For example, `pearsonr` and `spearmanr` do this. In other cases, like data has nans or there is an explicit argument like `nan_policy` that says what to do when there are nans in the data, a warning doesn't seem necessary. I think most functions in stats follow this format so I suggested adding a warning. If this sounds convincing to you, I guess there is no need for a third opinion. Otherwise, we can ask someone to take a look.

**Comment by mdhaber** [comment]
OK, it's up to you. Thanks for the review, and I trust your judgement on this.

**Review by tirthasheshpatel — APPROVED** [review]
I think this looks good! The note and test seem fine. Tests failures are unrelated. It would be nice to have this in 1.8.0. So, merging!

**Comment by tirthasheshpatel** [comment]
Thanks @mdhaber! And thanks @alexsherstinsky for your reviews!

**Comment by alexsherstinsky** [comment]
@tirthasheshpatel -- I am happy to do / help more as needed.  Please feel free to call upon me, and I will do my best.  In addition, @mdhaber and I started an effort a while ago, but the PR https://github.com/mdhaber/scipy/pull/60 was closed (in part because I was in a time crunch at work and also in part due to the fact that this work still had to wait for scipygh-1433 to be merged -- I do not know whether or not it was merged).  In any regard, I would be delighted to resume this.  Thank you very much.
